File: Chapter 09/C#/section_9_3_4_serverless_functions/Infrastructure/Aws/MockS3Event.cs
using System;
using System.Collections.Generic;
using System.Reactive.Linq;

namespace Chapter09.ServerlessFunctions.Infrastructure.Aws;

/// <summary>
/// THE AWS INFRASTRUCTURE CONTRACT (Reactive S3 Event):
/// 
/// DESIGN NOTE:
/// This record simulates the proprietary, strongly-typed object tree provided by the 
/// 'Amazon.Lambda.S3Events' NuGet package, but modeled as a Reactive Stream.
/// 
/// ARCHITECTURAL CRITIQUE:
/// In a reactive architecture, we treat cloud triggers as an infinite stream of incoming 
/// telemetry. However, notice the deep nesting required to find a simple filename: 
/// Records[0].S3.Object.Key. This is a textbook "Abstraction Leak." By forcing your 
/// business logic to navigate this specific hierarchy, the vendor has effectively 
/// "leaked" their internal data model into your application code. Your resizer logic 
/// is no longer a pure function; it is a downstream dependent of Amazon's design.
/// </summary>
public record MockS3Event(List<MockS3EventRecord> Records);
public record MockS3EventRecord(MockS3Entity S3);
public record MockS3Entity(MockS3Bucket Bucket, MockS3Object Object);
public record MockS3Bucket(string Name);
public record MockS3Object(string Key);

public static class AwsStreamFactory
{
    /// <summary>
    /// Wraps a static AWS event into a Reactive Observable stream.
    /// </summary>
    public static IObservable<MockS3Event> CreateS3Stream(string bucket, string file)
    {
        return Observable.Return(new MockS3Event(new List<MockS3EventRecord>
        {
            new MockS3EventRecord(new MockS3Entity(new MockS3Bucket(bucket), new MockS3Object(file)))
        }));
    }
}

File: Chapter 09/JS/section_9_3_4_serverless_functions/infrastructure/aws/mockS3Event.js
import { of } from 'rxjs';

/**
 * THE AWS INFRASTRUCTURE CONTRACT (Reactive S3 Event):
 * * DESIGN NOTE:
 * This class simulates the proprietary JSON structure that Amazon S3 sends 
 * to a Lambda handler, but wrapped in an Observable. 
 * * ARCHITECTURAL CRITIQUE:
 * In a reactive system, the cloud trigger is an infinite stream of events. 
 * Notice the nesting (Records -> s3 -> object). This forces the developer 
 * to write "Infrastructure-Aware" code. Even in a reactive pipeline, 
 * you are forced to pipe through Amazon's proprietary schema. This is an 
 * Abstraction Leak—the vendor's data structure is now a mandatory dependency 
 * in your reactive stream transformation.
 */
export class MockS3Event {
    /**
     * @param {string} bucketName 
     * @param {string} fileName 
     * @returns {Observable<Object>} A stream emitting the proprietary AWS JSON.
     */
    static create$(bucketName, fileName) {
        return of({
            Records: [{
                s3: {
                    bucket: { name: bucketName },
                    object: { key: fileName }
                }
            }]
        });
    }
}

/**
 * THE AWS RUNTIME CONTEXT (Reactive):
 * Simulates the environment stream provided by the AWS Lambda engine.
 */
export class MockLambdaContext {
    constructor() {
        this.awsRequestId = "aws-req-reactive-12345";
    }
}